crates/ruma-common/src/push/condition.rs

```diff
@@ -197,6 +211,35 @@ impl PushCondition {
                 .get(key)
                 .and_then(FlattenedJsonValue::as_array)
                 .is_some_and(|a| a.contains(value)),
+            #[cfg(feature = "unstable-msc4306")]
+            Self::ThreadSubscription { subscribed: must_be_subscribed } => {
+                let Some(has_thread_subscription_fn) = &context.has_thread_subscription_fn else {
+                    // If we don't have a function to check thread subscriptions, we can't
+                    // determine if the condition applies.
+                    return false;
+                };
+
+                // The event must have a relation of type `m.thread`.
+                if event.get_str("content.m\\.relates_to.rel_type") != Some("m.thread") {
+                    return false;
+                }
+
+                // Retrieve the thread root event ID.
+                let Some(Ok(thread_root)) = event
+                    .get_str("content.m\\.relates_to.event_id")
+                    .map(|str| str.parse::<OwnedEventId>())
+                else {
+                    return false;
+                };
+
+                let is_subscribed = has_thread_subscription_fn(&thread_root).await;
+
+                if *must_be_subscribed {
+                    is_subscribed
+                } else {
+                    !is_subscribed
+                }
```

Comment:
Is this really the correct logic? In particular `!*must_be_subscribed` implies "must not be subscribed" as opposed to "doesn't have to be subscribed"?

Reply:
Yeah, it is. The expected behavior is that notifications for a thread you've not subscribed to should *not* notify (unless some non-underride rule triggered before), so it implies "must not be subscribed to" in my understanding.

Reply:
I think that the semantics of "doesn't have to be subscribed" are equivalent to… no push condition at all :)

Reply:
I don't get it but don't have time to dive deeper right now.

Reply:
This condition is used either for:

- You **must** be subscribed to the thread
- You **must not** be subscribed to the thread

If we don't care whether the user is subscribed or not, we just don't use this condition. It is not a condition that can be used to check only if an event belongs to a thread, if that's were you are confused.

Reply:
Okay, that kinda makes sense, but... when is a push rule that triggers if a user *isn't* subscribed to a thread useful? Well, I guess the MSC would answer that.

Reply:
```suggestion
                *must_be_subscribed == is_subscribed
```

is the same thing. Is it easier or harder to read? Not sure.

Reply:
> Okay, that kinda makes sense, but... when is a push rule that triggers if a user _isn't_ subscribed to a thread useful? Well, I guess the MSC would answer that.

When you want messages in a thread to not trigger notifications in the room.

Reply:
> Okay, that kinda makes sense, but... when is a push rule that triggers if a user isn't subscribed to a thread useful? Well, I guess the MSC would answer that.

Yeah, it would short-circuit the evaluation of other underride rules (notably since it's before `m.rule.message`, you don't get a ping for _any_ message in a thread, for instance).